File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.py
import sys
input = sys.stdin.readline


def check(p):
    if p == "":
        return p
    left = 0
    right = 0
    is_correct = True
    for i in range(len(p)):
        if p[i] == "(":
            left += 1
        else:
            right += 1
        if left < right:
            is_correct = False
        if left == right:
            break
    u = p[:i+1]
    v = p[i+1:]

    if is_correct:
        return u + check(v)
    else:
        return "(" + check(v) + ")" + flip(u[1:-1])


def flip(s):
    return_value = ""
    for v in s:
        if v == "(":
            return_value += ")"
        else:
            return_value += "("
    return return_value


def solution(p):
    return check(p)


if __name__ == "__main__":
    print(solution("(()())()"))
    print(solution(")("))
    print(solution("()))((()"))

File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.java
class Solution {
    public String solution(String p) {
        if (p.isBlank()) {
            return "";
        }

        String[] split = splitString(p);
        String u = split[0];
        String v = split[1];

        if (isCorrect(u)) {
            return u + solution(v);
        }

        StringBuilder sb = new StringBuilder();
        sb.append("(");
        sb.append(solution(v));
        sb.append(")");

        for (int i = 1; i < u.length() - 1; i++) {
            if (u.charAt(i) == '(') {
                sb.append(")");
            } else {
                sb.append("(");
            }
        }

        return sb.toString();

    }

    public String[] splitString(String p) {
        int left = 0;
        int right = 0;
        for (int i = 0; i < p.length(); i++) {
            if (p.charAt(i) == '(') {
                left += 1;
            } else {
                right += 1;
            }

            if (left == right) {
                break;
            }
        }

        int index = left + right;
        return new String[]{p.substring(0, index), p.substring(index)};
    }

    public boolean isCorrect(String p) {
        int left = 0;
        int right = 0;
        for (int i = 0; i < p.length(); i++) {
            if (p.charAt(i) == '(') {
                left += 1;
            } else {
                right += 1;
            }

            if (right > left) {
                return false;
            }
        }

        return true;
    }
}